File: Java-Module/src/main/java/com/sim/javamodule/domain/product/domain/Product.java
package com.sim.javamodule.domain.product.domain;

import com.sim.javamodule.domain.user.domain.User;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
public class Product {
    @Id@GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "product_id")
    private Long id;

    private String name;
    private Long cost;
    private Long quantity;

    @ManyToOne(fetch = FetchType.LAZY)
    private User productUploader;

    public Product() {
    }

    public Product(String name, Long cost, Long quantity, User productUploader) {
        this.name = name;
        this.cost = cost;
        this.quantity = quantity;
        this.productUploader = productUploader;
    }

    public void sellProduct(Long amount){
        if(quantity<=0)
            throw new IllegalArgumentException("품절된 상품입니다.");
        if(quantity<amount)
            throw new IllegalArgumentException("최대 "+quantity+"개만큼 구매가능합니다.");
        this.quantity-=amount;
    }

    public void updateQuantity(Long amount){
        this.quantity+=amount;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Long getCost() {
        return cost;
    }

    public Long getQuantity() {
        return quantity;
    }

    public User getProductUploader() {
        return productUploader;
    }
}

File: Kotlin-Module/src/main/kotlin/com/sim/kotlin/domain/product/domain/Product.kt
package com.sim.kotlin.domain.product.domain

import com.sim.kotlin.domain.user.domain.User
import jakarta.persistence.*

/**
 * Kotlin에서는 getter, setter가 기본제공이다
 * <br> 하지만 setter만 막고싶을때는 어떻게 해야하지??
 */
@Entity
class Product(
    var name: String,
    var cost: Long,
    var quantity: Long,
    @ManyToOne(fetch = FetchType.LAZY) var productUploader: User
) {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "product_id")
    val id: Long? = null

    fun sellProduct(amount: Long){
        if(quantity<=0)
            throw IllegalArgumentException("품절된 상품입니다.")
        if(quantity<amount)
            throw IllegalArgumentException("최대 ${quantity}개만큼 구매가능합니다.")
        quantity-=amount
    }
}